libsolidity/ast/Types.h

```diff
@@ -881,6 +881,7 @@ class FunctionType: public Type
 	std::vector<std::string> const returnParameterTypeNames(bool _addDataLocation) const;
 	/// @returns the "self" parameter type for a bound function
 	TypePointer selfType() const;
+	virtual bool hasSelfType() const { return bound() && m_parameterTypes.size() > 0; }
```

Comment:
I think the above code in the typechecker is the place where it should be checked. How can you check during the `using` statement? The contract should be able to have non-bound functions too.